### PR TITLE
Adds DKIM support for multiple domains

### LINKED ISSUE #88: DKIM support for multiple domains
As discussed in #83, SMF needs a way to handle DKIM for multiple domains.

In particular, it should gracefully handle the domain defined in `$SMF_DOMAIN` for backward compatibility (going back to single DKIM), create DKIM public/private keypairs to be used for all source domains in `$SMF_CONFIG`, adapt `/var/db/dkim/` to a per-domain path style, and change `/etc/opendkim/opendkim.conf` to use `KeyTable`, `SigningTable`, `ExternalIgnoreList`, and `InternalHosts`.

Finally, SMF should check validity of all DKIM keys on startup.

This is underway.

### PR DESCRIPTION
**Pull request by dgraziotin**
Fixes #88 and merges #83.

Creates DKIM public/private keys and puts them into use.
    -  uses all domains defined in `$SMF_DOMAIN` and `$SMF_CONFIG` (only source domains for the latter) to generate public/private keypairs.
    - converts `/var/db/dkim/` to a per-domain path style `/var/db/dkim/domain.ext/` (if already present).
    - creates `KeyTable`, `SigningTable`, `ExternalIgnoreList`, and `InternalHosts` with all relevant values.
    - changes /etc/opendkim/opendkim.conf to use `KeyTable`, `SigningTable`, `ExternalIgnoreList`, and `InternalHosts`.
    - tests on SMF startup that all DNS entries are set correctly.

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -203,8 +203,56 @@ function start_postfix {
...
+        fi
+            chmod 400 /var/db/dkim/default.private
+            chown opendkim:opendkim /var/db/dkim/default.private
+            echo "Inserting ${HOSTNAME} data to /etc/opendkim/{KeyTable, SigningTable, TrustedHosts}"
+            echo "default._domainkey.${HOSTNAME} ${HOSTNAME}:default:/var/db/dkim/${HOSTNAME}/default.private
+            " >> /etc/opendkim/KeyTable
+
+            echo "${HOSTNAME} default._domainkey.${HOSTNAME}" >> /etc/opendkim/SigningTable
+
+            echo "${HOSTNAME}" >> /etc/opendkim/TrustedHosts
```
Comment (petslane) [inline entrypoint.sh:224]: Please fix the intention. And if you append with `>>`, does it mean that on every SMF restart new lines will be added to  `/etc/opendkim/{KeyTable, SigningTable, TrustedHosts}`?

File: entrypoint.sh [hunk 2]
```diff
@@ -203,8 +203,56 @@ function start_postfix {
...
+        
+        for virtualDomain in $virtualDomains; do
```
Comment (petslane) [inline entrypoint.sh:226]: `HOSTNAME` is probably also in `$virtualDomains`, so I suspect that configuration for `HOSTNAME` is done twice.

File: entrypoint.sh (hunk 1 above)
Comment (dgraziotin) [reply entrypoint.sh:223]: Thanks for checking! Indentation fixed at line 220, unless you meant something else.

The `>>` is required, because for all files I am inserting infos on multiple occasions:
-  first for the entry your PR creates (the ${HOSTNAME})
- over a loop for all `virtualDomain in $virtualDomains`

All new files in `/etc/opendkim/` are re-created at each `restart` or `down/up`, they are dynamic according to the inserted domains, so there is no danger of inserting new lines.
```
docker exec -it smf cat /etc/opendkim/KeyTable | wc -l
10
docker-compose restart
docker exec -it smf cat /etc/opendkim/KeyTable | wc -l
10
```

File: entrypoint.sh (hunk 2 above)
Comment (dgraziotin) [reply entrypoint.sh:225]: From what I understand from [here](https://github.com/dgraziotin/docker-simple-mail-forwarder/blob/master/entrypoint.sh#L68), `HOSTNAME` is not part of `$virtualDomains`. I did check for this empirically, to be sure. A folder `/var/db/dkim/${HOSTNAME}/default.private` was not created in a for loop over $virtualDomains. 

In any case, I think that this handles the case of reverting to a single domain (your PR), because HOSTNAME and all forwarded domains are kept separately.

File: entrypoint.sh (hunk 1 above)
Comment (petslane) [reply entrypoint.sh:223]: This was a good fix, same thing on line 234.

But I meant that the whole block of code from lines 216-223 and 231-242 is intented:
image

File: entrypoint.sh (hunk 2 above)
Comment (petslane) [reply entrypoint.sh:225]:
```
    - SMF_CONFIG=@example.com:xxx@gmail.com:pwd
    - SMF_DOMAIN=example.com
```
This is my config, based on https://github.com/dgraziotin/docker-simple-mail-forwarder/blob/99f27761e49e131a52b03159d69e7c652d25ffff/entrypoint.sh#L101 `virtualDomains` is a list of domains extracted from `emailFrom` which is the first element when split with `:`. I have to actually test but I'm quite sure that with my config, `example.com` is in `virtualDomains` and `SMF_DOMAIN`.

And what do you think, is #88 actually needed? Maybe dkim should always use a per-domain path style? Even if there is only one domain? This would mean removing [this](https://github.com/dgraziotin/docker-simple-mail-forwarder/blob/99f27761e49e131a52b03159d69e7c652d25ffff/entrypoint.sh#L194-L205) and removing  `SMF_DKIM_ALL` [here](https://github.com/dgraziotin/docker-simple-mail-forwarder/blob/99f27761e49e131a52b03159d69e7c652d25ffff/entrypoint.sh#L210).

File: entrypoint.sh (hunk 1 above)
Comment (dgraziotin) [reply entrypoint.sh]: Whops! Thanks for pointing it out. Fixed.

File: entrypoint.sh (hunk 2 above)
Comment (dgraziotin) [reply entrypoint.sh:224]: @petslane it was great that you pointed out your case, because it showcases my different use case! Luckily, both seem covered.

I think that we are both right: your `HOSTNAME` is part of `virtualDomains` because it is extracted from `SMF_CONFIG`. Normally people will have addresses `@${HOSTNAME}` in `SMF_CONFIG`. 

I do not. My `HOSTNAME` is dedicated to the SMTP server and does not appear in `SMF_CONFIG`. My domains use `HOSTNAME` in their MX entries. `HOSTNAME` is not used for sending or forwarding.

In light of this, I believe that both solutions should stay. #83 is needed for those with a single domain for everything (which I believe will be most people). #89 is for more complex cases. Everything is handled in `entrypoint.sh` and, if I did not make any mistake, things should stay clean all in all. @huan, what do you think?

Comment (petslane) [reply entrypoint.sh:224]: > In light of this, I believe that both solutions should stay. #83 is needed for those with a single domain for everything (which I believe will be most people). #89 is for more complex cases.

I have only one domain, why shouldn't I use your way of setting DKIM keys? From the user's point of view, it's currently complicated as there are 2 ways to set up DKIM, what option should the user choose? Why not have only one way to set up DKIM for SMF, your way, and there is no difference in how many domains are used (I expect that your way should work also with only 1 domain).

Comment (dgraziotin) [reply entrypoint.sh:224]: I would be OK to have both solutions merged, but I am not sure I could deliver this in short. PR to the PR are welcome, I guess 😄

**Comment by dgraziotin** [comment]
@petslane this last push merges #83 and #89 following my approach. You are right, it looks cleaner now. No config variable needed. Please review, though, and test. I hope I am covering all edge cases.

File: entrypoint.sh [hunk 3]
```diff
@@ -190,21 +190,70 @@ function start_postfix {
...
+    echo "${HOSTNAME} default._domainkey.${HOSTNAME}" >> /etc/opendkim/SigningTable
+    echo "${HOSTNAME}" >> /etc/opendkim/TrustedHosts
 
-        chmod 400 /var/db/dkim/default.private
-        chown opendkim:opendkim /var/db/dkim/default.private
+    for virtualDomain in $virtualDomains; do
+        # skip generating keys for $HOSTNAME twice in case it is also used as forwarded domain.
+        if [ "$virtualDomain" = "$HOSTNAME" ]; then
+            continue
+        fi
```
Comment (petslane) [inline entrypoint.sh:211]:
```suggestion
    allDomains="$virtualDomains"'
    [[ $virtualDomains =~ $HOSTNAME ]] || {
        virtualDomains="$virtualDomains $HOSTNAME"
    }
    for virtualDomain in $allDomains; do
```

**Comment by petslane** [comment]
I like this solution a lot better!

File: entrypoint.sh (hunk 3 above)
Comment (dgraziotin) [reply entrypoint.sh:211]: Very nice! And it does not manipulate `virtualDomains`, solving cases like mine.

File: test/simple-mail-forwarder.bats [hunk 4]
```diff
@@ -177,7 +177,9 @@
     if [[ "$SKIP_TEST" == *"DKIM"*  ]]; then
         skip "This test will fail on docker build workflow"
     fi
-    opendkim-testkey -d $SMF_DOMAIN -s default -vvv
-
+    cd /var/db/dkim/ && for domain in */ ; do
+        echo "Validating DKIM for ${domain::-1}"
+        opendkim-testkey -d ${domain::-1} -s default -vvv
```
Comment (petslane) [inline test/simple-mail-forwarder.bats:182]: This changes the current working directory.

```suggestion
    for domain in /var/db/dkim/*/ ; do
        echo "Validating DKIM for ${domain:13:-1}"
        opendkim-testkey -d ${domain:13:-1} -s default -vvv
```

Comment (dgraziotin) [reply test/simple-mail-forwarder.bats:182]: Fancy!

File: entrypoint.sh (hunk 3 above)
Comment (dgraziotin) [reply entrypoint.sh:211]: @petslane just to confirm, there is a typo on line 202 (extra `'`). I'm asking because, you never know with bash exoteric stuff.

Comment (petslane) [reply entrypoint.sh:211]: Yes, this is typo. `'` key too close to `enter` key.

**Comment by dgraziotin** [comment]
@petslane I am very satisfied of this PR now. Many thanks!

**Comment by dgraziotin** [comment]
@huan I think we are done here. I edited the PR description to reflect what's in the code right now. In the end it got way simpler than originally submitted. Code is complete, tests are complete. Feel free to merge or ask anything.

**Comment by huan** [comment]
@dgraziotin It's great to see this new feature has been implemented, and thank your effort for reviewing this PR to make it better @petslane 

Please help this PR to get at least one approvement before we have enough confidence to merge it.

**Comment by petslane** [comment]
Tried to test, but I get errors on startup:
```
forwarder_1  | >> ARGV arguments found. value:[start]
forwarder_1  | >> SMF_CONFIG found in ENV. use this settings for forward maps.
forwarder_1  | >> Setting password[xxx] for user @exaple.com ...
forwarder_1  | >> Setting password[xxx] for user yyy@yyy.yy ...
forwarder_1  | >> Setting password[xxx] for user yyy@yyy.yy ...
forwarder_1  | postmap: warning: /etc/postfix/virtual.db: duplicate entry: "@exaple.com
forwarder_1  | >> Set hostname to exaple.com
forwarder_1  | Migrating exaple.com keys to /var/db/dkim/exaple.com/
forwarder_1  | Inserting exaple.com data to /etc/opendkim/{KeyTable, SigningTable, TrustedHosts}
forwarder_1  | 'No such file or directory' messages might be displayed here. This is normal.
forwarder_1  | grep: /etc/opendkim/KeyTable: No such file or directory
forwarder_1  | grep: /etc/opendkim/SigningTable: No such file or directory
forwarder_1  | grep: /etc/opendkim/TrustedHosts: No such file or directory
forwarder_1  | OpenDKIM: this TXT record for petslane.eu should be present:
```

Looks like grep fails if these files are missing.

**Comment by petslane** [comment]
But then again, I can see that default keys were moved to domain-specific folders and https://www.mail-tester.com/ test gets 10/10.

Maybe `grep` shoule use `-s, --no-messages` argument? I think this should suppress "No such file or directory" errors.

**Comment by dgraziotin** [comment]
> Tried to test, but I get errors on startup:
> 
> ```
> forwarder_1  | >> ARGV arguments found. value:[start]
> forwarder_1  | >> SMF_CONFIG found in ENV. use this settings for forward maps.
> forwarder_1  | >> Setting password[xxx] for user @exaple.com ...
> forwarder_1  | >> Setting password[xxx] for user yyy@yyy.yy ...
> forwarder_1  | >> Setting password[xxx] for user yyy@yyy.yy ...
> forwarder_1  | postmap: warning: /etc/postfix/virtual.db: duplicate entry: "@exaple.com
> forwarder_1  | >> Set hostname to exaple.com
> forwarder_1  | Migrating exaple.com keys to /var/db/dkim/exaple.com/
> forwarder_1  | Inserting exaple.com data to /etc/opendkim/{KeyTable, SigningTable, TrustedHosts}
> forwarder_1  | 'No such file or directory' messages might be displayed here. This is normal.
> forwarder_1  | grep: /etc/opendkim/KeyTable: No such file or directory
> forwarder_1  | grep: /etc/opendkim/SigningTable: No such file or directory
> forwarder_1  | grep: /etc/opendkim/TrustedHosts: No such file or directory
> forwarder_1  | OpenDKIM: this TXT record for petslane.eu should be present:
> ```
> 
> Looks like grep fails if these files are missing.

Yeah, these aren't exactly errors. Mostly unwanted output. That's why I put ` 'No such file or directory' messages might be displayed here. This is normal.`. It's working, just ugly output. For some reason, the `-q` options of grep is not working. 



> But then again, I can see that default keys were moved to domain-specific folders and https://www.mail-tester.com/ test gets 10/10.
> 
> Maybe `grep` should use `-s, --no-messages` argument? I think this should suppress "No such file or directory" errors.

That was it, thanks. Was not aware of `-s`. Pushed.

**Comment by petslane** [comment]
No warnings anymore. LGTM.

**Comment by huan** [comment]
@petslane thank your very much for the testing, reviewing, and approvement!

Appreciate for all efforts from @dgraziotin and @petslane to make the SMF better, thank you very much!

P.S. I have bumped the version to 1.4 , the docker image should be published shortly.
